### PR TITLE
Feature/list view uikit

### LINKED ISSUE #2: Define overall flow of the APP
# Description
Create a design for each screen of the app, the screens should contain all of the components that are going to be used.

# Acceptance Criteria 
An image with the design of each screen

### PR DESCRIPTION
**Pull request by PeterChav01**
## Essential functionality implemented 

- Obtain value from searchbar and send it to the API
- Get data from API
- Display obtained data in the tableView
- Wait Async
- Alert added if an error occurs when fetching data from API
- Code Refactor

### CODE REVIEW

File: UIKit/ML App UIKit/Model/ProductData.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+/*
+struct Pokemon: Decodable {
+    var results: [PokemonDetail]
+}
+
+struct PokemonDetail: Decodable {
+    let name: String
+    let url: String
+}
+*/
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/Model/ProductData.swift]: Remove commented code

File: UIKit/ML App UIKit/NetworkManager.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    func fetchData(for url: String, completion: @escaping (Result<Product, Error>) -> Void)
+}
+
+class NetworkManager: ServiceProtocol {
+    func fetchData(for url: String, completion: @escaping (Result<Product, Error>) -> Void) {
+        if let url = URL(string: url) {
+            URLSession.shared.dataTask(with: .init(url: url)) { data, response, error in
+                if data != nil && error == nil {
+                    do {
+                        let fetchingData = try JSONDecoder().decode(Product.self, from: data!)
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/NetworkManager.swift]: Don't force unwrap

File: UIKit/ML App UIKit/ViewController.swift [hunk 3]
```diff
@@ -7,13 +7,220 @@
 
 import UIKit
 
-class ViewController: UIViewController {
+/// 1. ARC -  > Automatic Reference Counting
+/// 2. Strong v Weak v Unowned
+class ViewController: UIViewController, UITextFieldDelegate /* UISearchBarDelegate */ {
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: remove commented code

File: UIKit/ML App UIKit/ViewController.swift [hunk 4]
```diff
@@ -7,13 +7,220 @@
...
+    // MARK: - UI Components
+    
+    private lazy var searchBar: UISearchBar = {
+        let searchBar = UISearchBar()
+        searchBar.delegate = self
+        searchBar.translatesAutoresizingMaskIntoConstraints = false
+        return searchBar
+    }()
+    
+
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: extra space

File: UIKit/ML App UIKit/ViewController.swift [hunk 5]
```diff
@@ -7,13 +7,220 @@
...
+        let table = UITableView()
+        table.translatesAutoresizingMaskIntoConstraints = false
+        table.delegate = self
+        table.dataSource = self
+        return table
+    }()
+    
+    private let networkManager: ServiceProtocol
+
+    var product = Product(results: [])
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: this should be private

File: UIKit/ML App UIKit/ViewController.swift [hunk 6]
```diff
@@ -7,13 +7,220 @@
...
+    // MARK: - Init
+    init(networkMananager: ServiceProtocol) {
+        self.networkManager = networkMananager
+        super.init(nibName: nil, bundle: nil)
+        view.backgroundColor = .white
+    }
+    
+    deinit {
+        print("saliendo por la puerta")
+    }
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: deinit is not required

File: UIKit/ML App UIKit/ViewController.swift [hunk 7]
```diff
@@ -7,13 +7,220 @@
...
 
     override func viewDidLoad() {
         super.viewDidLoad()
-        // Do any additional setup after loading the view.
+        setupNavBar()
+        setupSearchBar()
+        setupTableView()
+        
+        hideKeyboardWhenTappedAround()
+
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: remove extra space

File: UIKit/ML App UIKit/ViewController.swift [hunk 8]
```diff
@@ -7,13 +7,220 @@
...
+
 
     override func viewDidLoad() {
         super.viewDidLoad()
-        // Do any additional setup after loading the view.
+        setupNavBar()
+        setupSearchBar()
+        setupTableView()
+
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: same here extra space

File: UIKit/ML App UIKit/ViewController.swift [hunk 9]
```diff
@@ -7,13 +7,220 @@
...
-        // Do any additional setup after loading the view.
+        setupNavBar()
+        setupSearchBar()
+        setupTableView()
+        
+        hideKeyboardWhenTappedAround()
+        
+    }
+    
+
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: extra space

File: UIKit/ML App UIKit/ViewController.swift [hunk 10]
```diff
@@ -7,13 +7,220 @@
...
+    
+    
+// MARK: - Functions
+    
+    private func setupNavBar() {
+        title = "MercadoLibre"
+    }
+    
+    private func setupSearchBar() {
+        // 1. Agregar search bar a la vista del viewcontroller
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: remove comment

File: UIKit/ML App UIKit/ViewController.swift [hunk 11]
```diff
@@ -7,13 +7,220 @@
...
+        view.addSubview(searchBar)
+        searchBar.placeholder = "Buscar Producto"
+        
+        // 2. Agregar constraints al search bar
+        // x, y, width, height
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: remove comments

File: UIKit/ML App UIKit/ViewController.swift [hunk 12]
```diff
@@ -7,13 +7,220 @@
...
+        let constraints: [NSLayoutConstraint] = [
+            searchBar.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor), // arriba
+            searchBar.leadingAnchor.constraint(equalTo: view.leadingAnchor), // izquierda
+            searchBar.trailingAnchor.constraint(equalTo: view.trailingAnchor) // derecha
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: same.. comments

File: UIKit/ML App UIKit/ViewController.swift [hunk 13]
```diff
@@ -7,13 +7,220 @@
...
+        ]
+        
+        NSLayoutConstraint.activate(constraints)
+    }
+    
+    private func setupTableView(){
+        view.addSubview(tableView)
+
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift:82]: remove extra space

File: UIKit/ML App UIKit/ViewController.swift [hunk 14]
```diff
@@ -7,13 +7,220 @@
...
+        DispatchQueue.main.async {
+            self.product = product
+            self.tableView.reloadData()
+        }
+    }
+    
+    func showAlert() {
+        let alert = UIAlertController(title: "Shit Happened", message: "Some Shit Happened", preferredStyle: .alert)
+        alert.addAction(UIAlertAction(title: "Well...Shit", style: .default, handler: nil))
+        self.present(alert, animated: true, completion: nil)
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift:104]: nice

File: UIKit/ML App UIKit/ViewController.swift [hunk 15]
```diff
@@ -7,13 +7,220 @@
...
+        return product.results.count
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = UITableViewCell()
+        cell.textLabel?.text = product.results[indexPath.row].title
+//        cell.textLabel?.text = product.results[indexPath.row].condition
+//        cell.textLabel?.text = String(product.results[indexPath.row].price)
+//        cell.textLabel?.text = product.results[indexPath.row].permalink
+//        cell.textLabel?.text = product.results[indexPath.row].thumbnail
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: remove prints

File: UIKit/ML App UIKit/ViewController.swift [hunk 16]
```diff
@@ -7,13 +7,220 @@
...
+        alert.addAction(UIAlertAction(title: "Well...Shit", style: .default, handler: nil))
+        self.present(alert, animated: true, completion: nil)
+    }
+    
+}
 
+// MARK: - UITableViewDelegate
+extension ViewController: UITableViewDelegate {
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+       // print(products[indexPath.row])
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: remove print

File: UIKit/ML App UIKit/ViewController.swift [hunk 17]
```diff
@@ -7,13 +7,220 @@
...
+//        cell.textLabel?.text = product.results[indexPath.row].thumbnail
+        return cell
+    }
 }
 
+
+//  MARK: - Get data from searchbar
+
+
+
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: extra spaces

File: UIKit/ML App UIKit/ViewController.swift [hunk 18]
```diff
@@ -7,13 +7,220 @@
...
+extension ViewController: UISearchBarDelegate {
+    
+    public func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
+        dispatchWorkItem?.cancel()
+        let workItem = DispatchWorkItem { [weak self] in
+            
+            print("\(searchText)")
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: remove print

File: UIKit/ML App UIKit/ViewController.swift [hunk 19]
```diff
@@ -7,13 +7,220 @@
...
+            self?.networkManager.fetchData(for: "https://api.mercadolibre.com/sites/MLM/search?q=\(searchText)") { result in
+                switch result {
+                case .success(let data): self?.reloadData(product: data)
+                case .failure(let error):
+                    print(error)
+                    print("error")
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: remove prints

File: UIKit/ML App UIKit/ViewController.swift [hunk 20]
```diff
@@ -7,13 +7,220 @@
...
+                    }
+                }
+            }
+        }
+        
+        dispatchWorkItem = workItem
+        if let waitTime = dispatchWorkItem {
+            queue.asyncAfter(deadline: .now() + 2.5, execute: waitTime)
+        }
+
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: remove extra space

File: UIKit/ML App UIKit/ViewController.swift [hunk 21]
```diff
@@ -7,13 +7,220 @@
...
+                    DispatchQueue.main.async {
+                        self?.showAlert()
+                    }
+                }
+            }
+        }
+        
+        dispatchWorkItem = workItem
+        if let waitTime = dispatchWorkItem {
+            queue.asyncAfter(deadline: .now() + 2.5, execute: waitTime)
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: constant 2.5 value should be moved to a private constants enum

File: UIKit/ML App UIKit/ViewController.swift [hunk 22]
```diff
@@ -7,13 +7,220 @@
...
+        }
+
+    }
+}
+
+// MARK: - Alert
+
+
+
+
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: ???

File: UIKit/ML App UIKit/ViewController.swift [hunk 23]
```diff
@@ -7,13 +7,220 @@
...
+//    
+//    @objc func navigateToViewController() {
+//        show(
+//            ViewController(
+//                networkMananager: NetworkManager()
+//            ),
+//            sender: nil
+//        )
+//    }
+//}
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: remove

File: UIKit/ML App UIKit/ViewController.swift [hunk 24]
```diff
@@ -7,13 +7,220 @@
...
+
+extension UIViewController {
+    func hideKeyboardWhenTappedAround() {
+        let tap: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(UIViewController.dismissKeyboard))
+        tap.cancelsTouchesInView = false
+        view.addGestureRecognizer(tap)
+    }
+
+    @objc func dismissKeyboard() {
+        view.endEditing(true)
```
Comment (AlejandroVC67) [inline UIKit/ML App UIKit/ViewController.swift]: this should be on the class definition, it should not be on a separate extension ..also private functions

**Review by AlejandroVC67 — CHANGES_REQUESTED** [review]

**Comment by PeterChav01** [comment]
code updated according to the observations

**Review by AlejandroVillaCardenas — APPROVED** [review]
Good job!

**Review by AlejandroVC67 — APPROVED** [review]
LGTM

**Pull request merged**